python_modules/dagster/dagster/_core/definitions/asset_spec.py

```diff
@@ -0,0 +1,117 @@
+from typing import AbstractSet, Any, Iterable, Mapping, NamedTuple, Optional, Union
+
+import dagster._check as check
+from dagster._annotations import PublicAttr
+from dagster._core.definitions.assets import AssetsDefinition
+from dagster._core.definitions.backfill_policy import BackfillPolicy
+from dagster._core.definitions.source_asset import SourceAsset
+
+from .auto_materialize_policy import AutoMaterializePolicy
+from .events import (
+    AssetKey,
+    CoercibleToAssetKey,
+)
+from .freshness_policy import FreshnessPolicy
+from .metadata import MetadataUserInput
+
+
+class AssetSpec(
+    NamedTuple(
+        "_AssetSpec",
+        [
+            ("asset_key", PublicAttr[AssetKey]),
+            ("deps", PublicAttr[AbstractSet[AssetKey]]),
```

Comment:
The behavior for multi-assets currently is that, if the `internal_asset_deps` parameter is not provided, every output asset is assumed to depend on every input asset.

Is my understanding correct that here we'd be requiring users to specify this all-to-all dependency on their own?

I assume this means that if you're using asset specs then you can't also specify the `deps` param of `@multi_asset`? What about if there are parameter inputs? Are those allowed, and, if so, how do those get hooked up specs?

Reply:
ill get some tests written up and try to flesh this out

Reply:
alright, PR updated with the idea that `AssetSpec.deps` is the **only** way to specify dependencies 

> can't also specify the deps param of @multi_asset? 

correct. The `AssetSpec.deps` that are not internal to the `@multi_asset` form the "deps" except when...

>  What about if there are parameter inputs? Are those allowed, and, if so, how do those get hooked up specs?

you can have parameter inputs, but there must be a corresponding `AssetSpec.deps` entry for it.

Reply:
Got it - those semantics make sense to me. A little worried that "updated with the idea that AssetSpec.deps is the only way to specify dependencies" will be too cumbersome, but I don't have a great alternative.

Reply:
Ya same. We can always relax constraints in the future if we figure out something we like.